lib/Optimizer/Transforms/AssignIDs.cpp

```diff
@@ -0,0 +1,171 @@
+    if (alloc->hasAttr("qid"))
+      return failure();
+
+    auto qid = (*counter)++;
+
+    rewriter.startRootUpdate(alloc);
+    alloc->setAttr("qid", rewriter.getUI32IntegerAttr(qid));
+    rewriter.finalizeRootUpdate(alloc);
+
+    return success();
+  }
+};
+
+std::optional<uint> findQid(Value v) {
+  auto defop = v.getDefiningOp();
+  if (!defop)
+    return std::nullopt;
+
+  if (!isa<quake::WireType>(v.getType()))
+    return std::nullopt;
+
+  if (!quake::isLinearValueForm(defop))
+    defop->emitOpError("assign-ids requires operations to be in value form");
+
+  if (isBeginOp(defop)) {
+    assert(defop->hasAttr("qid") && "qid not present for beginOp");
+    uint qid = defop->getAttr("qid").cast<IntegerAttr>().getUInt();
+    return std::optional<uint>(qid);
+  }
+
+  // Figure out matching operand
+  size_t i = 0;
+  for (; i < defop->getNumResults(); i++)
+    if (defop->getResult(i) == v)
+      break;
+
+  // Special cases where result # != operand #
+  if (isMeasureOp(defop))
+    i = 0;
+  else if (hasClassicalInput(defop))
+    i++;
+
+  return findQid(defop->getOperand(i));
+}
+
+class SinkOpPat : public OpRewritePattern<quake::SinkOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  using Base = OpRewritePattern<quake::SinkOp>;
+
+  SinkOpPat(MLIRContext *context) : OpRewritePattern<quake::SinkOp>(context) {}
+
+  LogicalResult matchAndRewrite(quake::SinkOp release,
+                                PatternRewriter &rewriter) const override {
+    auto qid = findQid(release.getOperand());
+
+    if (!qid.has_value())
+      release->emitOpError(
+          "Corresponding null_wire not found for sink, illegal ops present");
+
+    rewriter.startRootUpdate(release);
+    release->setAttr("qid", rewriter.getUI32IntegerAttr(qid.value()));
+    rewriter.finalizeRootUpdate(release);
+
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Pass implementation
+//===----------------------------------------------------------------------===//
+
+struct AssignIDsPass : public cudaq::opt::impl::AssignIDsBase<AssignIDsPass> {
+  using AssignIDsBase::AssignIDsBase;
+
+  void runOnOperation() override {
+    auto func = getOperation();
+
+    // Blocks will cause problems for assign-ids, ensure there's only one
+    if (func.getBlocks().size() != 1) {
+      func.emitOpError("multiple blocks not currently supported in assign-ids");
+      signalPassFailure();
+      return;
+    }
+
+    assign();
+  }
+
+  void assign() {
+    auto *ctx = &getContext();
+    func::FuncOp func = getOperation();
+    RewritePatternSet patterns(ctx);
+    unsigned x = 0;
+    patterns.insert<NullWirePat>(ctx, &x);
+    patterns.insert<SinkOpPat>(ctx);
+    ConversionTarget target(*ctx);
+    target.addLegalDialect<quake::QuakeDialect>();
+    target.addDynamicallyLegalOp<quake::NullWireOp>(
+        [&](quake::NullWireOp alloc) { return alloc->hasAttr("qid"); });
```

Comment:
The attribute approach is brittle as other transformations will not necessarily keep the information. We should really use WireSet, BorrowWire, ReturnWire for explicitly mapping the identity of a qubit from an infinite space (virtual registers) to a finite space (physical registers).

Reply:
Makes sense, right now Assign IDs and Dependency Analysis are very tangled passes, so they pretty much have to be run one right after another, and Dependency Analysis has to check that the attributes are present.